// hubot-slack/src/reaction-message.js
// Generated by CoffeeScript 1.12.6
(function() {
  var Message, ReactionMessage,
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  Message = require('hubot').Message;

  ReactionMessage = (function(superClass) {
    extend(ReactionMessage, superClass);

    function ReactionMessage(type, user, reaction, item_user, item, event_ts) {
      this.type = type;
      this.user = user;
      this.reaction = reaction;
      this.item_user = item_user;
      this.item = item;
      this.event_ts = event_ts;
      ReactionMessage.__super__.constructor.call(this, this.user);
      this.type = this.type.replace('reaction_', '');
    }

    return ReactionMessage;

  })(Message);

  module.exports = ReactionMessage;

}).call(this);
