// imubot/index.js
// Generated by CoffeeScript 1.6.3
(function() {
  var Adapter, Brain, CatchAllMessage, EnterMessage, LeaveMessage, Listener, Message, Response, Bot, TextListener, TextMessage, TopicMessage, User, _ref, _ref1;

  User = require('./src/user');

  Brain = require('./src/brain');

  Bot = require('./src/bot');

  Adapter = require('./src/adapter');

  Response = require('./src/response');

  _ref = require('./src/listener'), Listener = _ref.Listener, TextListener = _ref.TextListener;

  _ref1 = require('./src/message'), Message = _ref1.Message, TextMessage = _ref1.TextMessage, EnterMessage = _ref1.EnterMessage, LeaveMessage = _ref1.LeaveMessage, TopicMessage = _ref1.TopicMessage, CatchAllMessage = _ref1.CatchAllMessage;

  module.exports = {
    User: User,
    Brain: Brain,
    Bot: Bot,
    Adapter: Adapter,
    Response: Response,
    Listener: Listener,
    TextListener: TextListener,
    Message: Message,
    TextMessage: TextMessage,
    EnterMessage: EnterMessage,
    LeaveMessage: LeaveMessage,
    TopicMessage: TopicMessage,
    CatchAllMessage: CatchAllMessage
  };

  module.exports.loadBot = function(adapterPath, adapterName, enableHttpd, botName, botAlias) {
    return new Bot(adapterPath, adapterName, enableHttpd, botName, botAlias);
  };

}).call(this);
